packages/vitest/src/types/coverage.ts

```diff
@@ -52,14 +52,11 @@ export interface CoverageProviderModule {
   stopCoverage?(): unknown | Promise<unknown>
 }
 
-export type CoverageReporter = keyof ReportOptions
-
-type CoverageReporterWithOptions<ReporterName extends CoverageReporter = CoverageReporter> =
-   ReporterName extends CoverageReporter
-     ? ReportOptions[ReporterName] extends never
-       ? [ReporterName, {}] // E.g. the "none" reporter
-       : [ReporterName, Partial<ReportOptions[ReporterName]>]
-     : never
+type CoverageReporterWithOptions = {
+  [key in keyof ReportOptions]: [key, ReportOptions[key] extends never ? {} : Partial<ReportOptions[key]>]
+}[keyof ReportOptions] | [string, Record<string, any>]
+
+export type CoverageReporter = CoverageReporterWithOptions[0]
```

Comment:
Do these changes still pass the type checking tests in `test/coverage-test`? This part was quite tricky to get right.

Run `pnpm test:types` in `vitest/test/coverage-test`.

Reply:
This one is passing:
```ts
type CoverageReporterWithOptions = {
  [key in keyof ReportOptions]: [key, ReportOptions[key] extends never ? {} : Partial<ReportOptions[key]>]
}[keyof ReportOptions]
```
But when we add ` | [string, Record<string, any>]` it's not.
Unfortunately  I'm not sure if it's possible to solve because typescript doesn't have "string except ..." type.
![image](https://github.com/vitest-dev/vitest/assets/67923777/4373538d-2527-4fae-b5d7-7b88a7c86f44)
This is happening because `['html', { blockName: 'string' }]` is assignable to `[string, Record<string, any>]` which should be a type **only** for custom reporter.

However this one looks promising:
```ts
type CoverageReporterWithOptions = {
  [key in keyof ReportOptions]: [key, ReportOptions[key] extends never ? {} : Partial<ReportOptions[key]>]
}[keyof ReportOptions] | [string, Record<string, never>]
```
With this one we cannot pass custom properties to custom reporters but still, it's the best solution.
![image](https://github.com/vitest-dev/vitest/assets/67923777/1344114b-6b7e-422f-86bb-f10cd6a3b85c)
Tests only complains about unknown reporters - but it's expected behavior, I mean tests should be adjusted.

What you think?